### ISSUE TITLE
Gradient of QR fails

### ISSUE THREAD

**Issue by dpfau** [google/jax#2863]
There seems to be a major bug in the gradient of the QR decomposition. The forward-mode derivative is accurate for Q but very far off for R. There also seems to be no check in the tests making sure the gradient is accurate.

This reproduces the error:

```
import jax
import jax.numpy as jnp
import numpy as np

x = np.random.randn(3, 3)
dx = np.random.randn(3, 3)

primals, tangents = jax.jvp(jnp.linalg.qr, (x,), (dx,))
q, r = primals
dq, dr = tangents

dt = 1e-6
dq_ = (np.linalg.qr(x + dt * dx)[0] - np.linalg.qr(x)[0]) / dt
dr_ = (np.linalg.qr(x + dt * dx)[1] - np.linalg.qr(x)[1]) / dt

assert jnp.allclose(x, q @ r, atol=1e-5, rtol=1e-5)  # passes
assert jnp.allclose(dq, dq_, atol=1e-5, rtol=1e-5)  # passes
assert jnp.allclose(dx, q @ dr_ + dq_ @ r, atol=1e-5, rtol=1e-5)  # passes
assert jnp.allclose(dr, dr_, atol=1e-5, rtol=1e-5)  # fails
```

**Comment by dpfau** [comment]
Oddly, it seems to be accurate for the rightmost column but wrong everywhere else. Here is one output of a run of the above script:

The JAX computation:
```
[[-1.5398827  -0.64605635  0.83935267]
 [ 0.          0.0303016  -0.43336833]
 [ 0.          0.         -0.9261702 ]]
```

The finite difference computation:
```
[[-0.89484418  0.61289863  0.83935263]
 [ 0.         -0.42105111 -0.43336874]
 [ 0.          0.         -0.92617024]]
```

**Comment by dpfau** [comment]
Actually I take it back...there is a call to `jtu.check_jvp` in `testQr` in `linalg_test.py`. Not sure why it's not failing.

**Comment by dpfau** [comment]
Also odd...when I copy the implementation of `qr_jvp_rule` line by line and run it myself, it works fine. I suspect there is something subtle going on with one of the backends that's not getting caught by the tests.

**Comment by dpfau** [comment]
This is on the CPU backend, both on my laptop and in Colab.

**Comment by hawkinsp** [comment]
@j-towns

**Comment by hawkinsp** [comment]
I note that our test actually does fail, we just get "lucky" with the random numbers it generates as input. If I call the rng twice it fails in several test cases. @j-towns wrote the QR JVP and wrote up his derivation; Jamie: would you mind taking a look?

**Comment by j-towns** [comment]
Interesting, I'll try to take a look at this today. FYI the derivation I wrote is [here](https://j-towns.github.io/papers/qr-derivative.pdf).

**Comment by dpfau** [comment]
Thanks. Yes, I saw the derivation, and checked it manually. The problem is
perhaps deeper than the derivation itself and has more to do with something
weird happening in the back end if the exact same function implemented as a
jvp rule for a primitive gives different results than executing the bare
function.

**Comment by j-towns** [comment]
I believe I've fixed the issue. Can you check that this fix in the above pr works for you @dpfau?

**Comment by dpfau** [comment]
Yes, that works for me. Any idea what was going wrong?

**Comment by j-towns** [comment]
Hmm this is a confusing situation. So the jvp rule we write needs to apply to lax_linalg.qr_p, which is not exactly the qr decomposition. It returns a pair `q, r_` where `r_` is not triangular (it's a full matrix), but the upper triangle of `r_` is equal to the `r` from the true qr decomposition. So we could call lax_linalg.qr_p a 'pseudo-qr decomposition'.

In order to define a correct derivative (so that the tests pass) we need a jvp rule for this pseudo-qr decomposition. I'm not sure what that should be because I'm not sure where the values in the lower-triangle of lax_linalg.qr_p come from. Maybe a better solution would be to redefine lax_linalg.qr_p so that it selects out the lower triangle (i.e. make it into a valid qr decomposition).

You can see the higher level lax.qr (_not_ lax.qr_p) selecting out the upper triangle on [this line](https://github.com/google/jax/blob/master/jax/lax_linalg.py#L66).

**Comment by j-towns** [comment]
```python

In [2]: import jax.lax_linalg as ll

In [3]: q, r_ = ll.qr_p.bind(x, False)

In [4]: r_

# Junk in lower triangle of r_
Out[4]: 
DeviceArray([[ 1.3606681 ,  2.2774568 ,  1.4783324 ],
             [-0.4206181 ,  0.6496488 ,  0.62955266],
             [-0.02333226,  0.5480194 , -0.8703172 ]], dtype=float32)
```

**Comment by j-towns** [comment]
Actually an easy fix might be to give `lax.qr` (the high-level, correct qr decomposition) a custom_jvp. I'll try that...

**Comment by hawkinsp** [comment]
It's not actually junk. On CPU and GPU, it's an encoding of the elementary Householder reflectors (i.e., matrix `q`): http://www.netlib.org/lapack/explore-html/df/dc5/group__variants_g_ecomputational_ga3766ea903391b5cf9008132f7440ec7b.html
On TPU, it will always be zero.

I suspect the correct fix here is to push the `np.triu` down into the implementation of the CPU and GPU translation rules.

(Alternatively, perhaps the QR primitive should return the packed representation, per https://github.com/google/jax/issues/2322 but I think that's most likely something that we should consider separately.)

**Comment by j-towns** [comment]
OK cool. I've pushed the custom_jvp thing to the pr, not sure if that's suitable as an interrim solution. FYI it looks like the tests still sometimes fail for complex inputs. I'm guessing that's because the jvp implementation is incorrect in that case (I'm not sure why). Should we for now raise a NotImplemented for complex inputs and not test that case?

**Comment by hawkinsp** [comment]
I agree that raising `NotImplemented` is preferable to a wrong output. But if you had ideas on how to fix the complex JVP, that'd be awesome too.

**Comment by j-towns** [comment]
> But if you had ideas on how to fix the complex JVP, that'd be awesome too.

Having a look at that, but will submit a separate pr if I fix it.

**Comment by j-towns** [comment]
I think the false positive test passes might have been caused by identical draws from np.linalg.RandomState(0), which seems to be used as PRNG in both test_util.rand_default and test_util.check_jvp. Gonna push something to the above pr to fix that...

**Comment by hawkinsp** [comment]
Yes, I agree with the last comment. We should not be choosing such closely correlated test cases, and I suspect the fallout from this will be wider than just QR.

**Comment by hawkinsp** [comment]
https://github.com/google/jax/pull/2870 pushes the `triu` into the translation rule and disables the complex case. Should I submit that first and then you submit the complex input fix?

**Comment by hawkinsp** [comment]
I'll also look into the PRNG use in test cases, since I suspect it's a bigger job.

**Comment by j-towns** [comment]
>#2870 pushes the triu into the translation rule and disables the complex case. Should I submit that first and then you submit the complex input fix?

Yes, that makes sense.

**Comment by dpfau** [comment]
Thanks! And glad this was able to flush out the testing rng seed issue!
